velox/connectors/hive/storage_adapters/hdfs/HdfsReadFile.cpp

```diff
@@ -30,37 +69,35 @@ HdfsReadFile::HdfsReadFile(hdfsFS hdfs, const std::string_view path)
       hdfsGetLastError());
 }
 
+HdfsReadFile::~HdfsReadFile() {
+  // should call hdfsFreeFileInfo to avoid memory leak
+  hdfsFreeFileInfo(fileInfo_, 1);
+}
+
 void HdfsReadFile::preadInternal(uint64_t offset, uint64_t length, char* pos)
     const {
   checkFileReadParameters(offset, length);
-  auto file = hdfsOpenFile(hdfsClient_, filePath_.data(), O_RDONLY, 0, 0, 0);
-  VELOX_CHECK_NOT_NULL(
-      file,
-      "Unable to open file {}. got error: {}",
-      filePath_,
-      hdfsGetLastError());
-  seekToPosition(file, offset);
-  uint64_t totalBytesRead = 0;
-  while (totalBytesRead < length) {
-    auto bytesRead = hdfsRead(hdfsClient_, file, pos, length - totalBytesRead);
-    VELOX_CHECK(bytesRead >= 0, "Read failure in HDFSReadFile::preadInternal.")
-    totalBytesRead += bytesRead;
-    pos += bytesRead;
-  }
-
-  if (hdfsCloseFile(hdfsClient_, file) == -1) {
-    LOG(ERROR) << "Unable to close file, errno: " << errno;
+  static thread_local HdfsFile file(hdfsClient_);
```

Comment:
You need instance thread local instead of global thread local here.  Check out https://github.com/facebook/folly/blob/main/folly/ThreadLocal.h

Reply:
Done.